### ISSUE TITLE
Proposal: `winget run`, a way to run a package's canonical executable that winget-cli controls

### ISSUE THREAD

**Issue by cutecycle** [microsoft/winget-cli#2280]
### Description of the new feature / enhancement

Many winget installations don't automatically add to path, so many programs that people expect to immediately be able to run from the cli after installing (especially with `-h`) are `Not Found`. 

I think this creates a disconnect from most people's expectations for a package manager; and independently of opinion about intuition from folks coming from the linux (or `choco`) world, adds a hit to the productivity of `winget` users. 

It is understandably hard to corral every MSI maintainer into enabling/adding Add To PATH functionality, especially for GUI applications; so maybe ideally, a solution to this just doesn't even try that. 

# Precedent: `npx`

`npm`'s `npx` is sort of this idea, but takes it a bit further than I'm thinking: it wraps and sandboxes the entire `install -g` process and runs a CLI tool on-the-fly. 

# Interface 

````
PS > winget install -h Vim.Vim
PS > vim

The term 'vim' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

PS > winget run vim

* vim opens*
```

# Precedence system 

### Proposed technical implementation details

My understanding of `winget install` is that when installing, it has an order of precedence for resolving a package identifier, detailed in: 

https://github.com/microsoft/winget-cli/blob/master/src/AppInstallerCLICore/Commands/InstallCommand.cpp#L62
https://github.com/microsoft/winget-cli/blob/master/src/AppInstallerCLICore/Workflows/CompletionFlow.cpp#L124
.


The same system could be used for resolving the target package of `winget run`.

I'm not as knowledgeable about the concept of the "primary executable" of an MSI, given there can be many executables in the output directory, and it may be somewhat of a nebulous concept (especially when some packages install .exes with `name-version-date.exe`, others may install a "dx9, dx11, dx12", some may install an "avx2, sse4", some may install an "x86, x64", etc. 

Maybe:
* above all, a new installer manifest yml variable, `Installers.Installer.RunExecutablePath` takes precedence. This would be the only allowed system for `winget run --silent`, for CLI stability/security and as a "carrot" for admins who want to implement `winget` at scale. Even though it's still manual, it provides a direct and central location `winget-pkgs` for this, rather than having to reach out to every package maintainer in every repository individually 

* list executable files in a directory. 
* filter remove names "install", "uninstall".
* if the end result is a single executable, run that.
* if not confident (two executables), and allow user to choose from a list, and set it permanently in their `AppData`.

Sorry if this proposal sounds silly!

**Comment by duggaraju** [comment]
vim installer has a "Create .bat files" options which create the .bat files in PATH  (C:\windows) that can launch vim/gvim.   Looks like winget install doesn't set that option

**Comment by wmstack** [comment]
I suggest that this feature is actually pretty useful. Cargo is technically a *package* manager as well, yet it has a command for running stuff that isn't in the path. Even NPM adds those binaries which are not in the path when running tasks with `npm run`. It seems like a good thing overall.

I'd dare say most users don't know internally where winget installs the binary. I believe it is in Appdata Local, but it isn't explicit. Plus, it is actually awful user experience if, after installing a package, that you have to go look in the file explorer for the binary before you can run it. So if you install something with `winget`, it makes sense to be able to run it with `winget`.

I also want to mention Cargo binutils, which I know to have moved on to this kind of option. If I understand correctly, in the past the cargo binutils crate had you specify which artifact to inspect, and now it automatically finds the artifact for you. That is pretty helpful, because users rarely have to find the binary in the `target/` directory as well.

**Comment by cutecycle** [comment]
@duggaraju : yeah; I think in the short term, getting the MSI to set the vim installer's PATH option makes sense!

Though what I'm thinking is... there are hundreds of `vim`s out there, especially if winget's method is to use MSIs pulled directly from maintainers.

@wmstack : I think it could be big for UX! and yeah, I always think about how when I first started using node, `npm i -g` was so good that on windows/mac machines I found myself looking for CLI tools on npm instead of homebrew/macports/choco/sourceforge or what have you.


@denelon : true; I think the common expectation with OSS package managers is something like, you call the executable by its moniker directly, right after installing. Maybe, winget can add a directory to PATH creates an alias/shortcut for `Package`, or `moniker`, that maps to `winget run Package`, and then fork from there (though it may make the process lifecycle look weird to some folks...):

```
> winget install Vim.Vim

> vim 
Vim does not have a default PATH configuration. Would you like winget to open: 
[1] Vim.Vim / gvim.exe?
[2] Vim.Vim / vim.exe?

> 2

> Would you like this to be the default method for opening Vim.Vim? [Y]es / [N]o

> <launch>
```


automation (I'm not actually sure if `jq` is in winget, yay `ConvertFrom-Json`, but for example): 

```
> winget install -h jq --winget-controls-path --default-command 'jq.exe' 
jq installed. 
$path/jq.exe set as default executable for the jq command.

```

**Comment by denelon** [comment]
Related to:
* #2252

We're looking to add a default path and executable to the manifest which might be part of the key to a deterministic way to unlock some of this.
